addons/product/models/product.py

```diff
@@ -784,3 +776,29 @@ def get_import_templates(self):
             'label': _('Import Template for Vendor Pricelists'),
             'template': '/product/static/xls/product_supplierinfo.xls'
         }]
+
+    @api.model
+    def create(self, vals):
+        if not vals.get("uom_purchase_id") and vals.get("product_tmpl_id"):
+            template = self.env["product.template"].browse(vals.get("product_tmpl_id"))
+            vals["uom_purchase_id"] = template.uom_id.id
+        return super().create(vals)
+
+    @api.onchange("product_tmpl_id")
+    def _onchange_product_tmpl_id(self):
+        self.uom_purchase_id = self.product_tmpl_id.uom_id
+
+    @api.constrains("uom_purchase_id")
+    def _check_uom(self):
+        if any(supplierinfo.product_uom_id and supplierinfo.uom_purchase_id and supplierinfo.product_uom_id.category_id != supplierinfo.uom_purchase_id.category_id for supplierinfo in self):
+            raise ValidationError(_('The default Unit of Measure and the purchase Unit of Measure must be in the same category.'))
+
+    @api.constrains("uom_purchase_id", "name")
+    def _check_same_uom_purchase_per_supplier(self):
+        for seller in self:
+            uoms = self.search([
+                ("name", "=", seller.name.id),
+                ("product_tmpl_id", "=", seller.product_tmpl_id.id)
+            ]).mapped("uom_purchase_id")
+            if len(uoms) > 1:
+                raise ValidationError(_('You can not affect various Purchase Unit of Measure for the same supplier.'))
```

Comment:
Can't we have same Vendor Selling 1 unit at 1$ and 1 Dozen at 10$ ?

Reply:
It's a use case that is not allowed by this new design. 

I'm not sure It's real use case. I implemented Odoo purchase in a quite complex database : I faced many times the situation with supplier 1 -> UoM A, supplier 2 -> UoM B. (for eggs, for 25kg bag of flour, etc...) but never a supplier1 -> UoM A and B for the same product. Do you have exemple ?

First, I tried to implement it but it's harder. So for the time being, I blocked it, but we can imagine to implement in a second time. (doesn't require database structure changes, only more code)